12-spatial-cv.Rmd

```diff
@@ -315,184 +309,196 @@ This is because there are still a few points in the test and training data which
-                      measures = mlr::auc)
+# reduce verbosity
+lgr::get_logger("mlr3")$set_threshold("warn")
+# run spatial cross-validation
+sp_cv = resample(task = task,
+                 learner = learner,
+                 resampling = perf_level)
 ```
 
-<!-- sp_cv and conv_cv have been saved in spatialcv.Rdata. I needed to run the modeling outside of the book since knitr sets its own seed and I am not sure if this actually helps to make sure that the same partitions are used in the cv.
-I really don't understand why I have to load spatialcv.Rdata here a third time...-->
-
-The output of the preceding code chunk is a bias-reduced assessment of the model's predictive performance, as illustrated in the following code chunk  (required input data is saved in the file `spatialcv.Rdata` in the book's GitHub repo):
+The output of the preceding code chunk is a bias-reduced assessment of the model's predictive performance, as illustrated in the following code chunk (required input data is saved in the file `12-sp_cv.rds` in the book's GitHub repo).
+As performance measure, we again choose the AUROC.
 
-```{r 11-spatial-cv-21, echo=FALSE}
-sp_cv = readRDS("extdata/sp_cv.rds")
-conv_cv = readRDS("extdata/conv_cv.rds")
+```{r 12-spatial-cv-21, echo=FALSE}
+bm_agg = readRDS("extdata/12-sp_conv_cv.rds")
+sp_cv = bm_agg$resample_result[[1]]
 ```
 
-```{r 11-spatial-cv-22}
+```{r 12-spatial-cv-22}
 # summary statistics of the 500 models
-summary(sp_cv$measures.test$auc)
+sp_cv$score(measures = msr("classif.auc"))[, summary(classif.auc)]
 # mean AUROC of the 500 models
-mean(sp_cv$measures.test$auc)
+sp_cv$aggregate(measures = msr("classif.auc"))
 ```
 
 To put these results in perspective, let us compare them with AUROC\index{AUROC} values from a 100-repeated 5-fold non-spatial cross-validation (Figure \@ref(fig:boxplot-cv); the code for the non-spatial cross-validation\index{cross-validation} is not shown here but will be explored in the exercise section).
 As expected, the spatially cross-validated result yields lower AUROC values on average than the conventional cross-validation approach, underlining the over-optimistic predictive performance due to spatial autocorrelation\index{autocorrelation!spatial} of the latter.
 
 ```{r boxplot-cv, echo=FALSE, fig.width=6, fig.height=9, fig.cap="Boxplot showing the difference in AUROC values between spatial and conventional 100-repeated 5-fold cross-validation.", fig.scap="Boxplot showing AUROC values."}
-# Visualization of non-spatial overfitting
-boxplot(sp_cv$measures.test$auc,
-        conv_cv$measures.test$auc,
-        col = c("lightblue2", "mistyrose2"),
-        names = c("spatial CV", "conventional CV"), 
-        ylab = "AUROC")
+library(ggplot2)
+# extract the AUROC values and put them into one data.table
+d = purrr::map_dfr(bm_agg$resample_result, ~ .$score(msr("classif.auc")))
+# rename the levels of resampling_id
+d[, resampling_id := as.factor(resampling_id) |>
```

Comment:
Don't fully understand this `:=` syntax. Know `echo=FALSE` but what is it out of interest?

Reply:
That is **data.table**'s update-by-reference syntax. In this case, we convert the resampling_id column into a factor. The corresponding dplyr syntatx would be `d = mutate(d, resampling_id = as.factor(resampling_id))`